llvm/tools/sycl-post-link/CUDASpecConstantToSymbol.cpp

```diff
@@ -0,0 +1,267 @@
+    if (!std::get<0>(Res))
+      continue;
+    if (std::get<1>(Res)->getName() == KernelName)
+      return std::get<0>(Res);
+  }
+
+  return nullptr;
+}
+
+void CUDASpecConstantToSymbolPass::setUpPlaceholderEntries(Module &M) {
+  NamedMDNode *AnnotationMD = M.getNamedMetadata("nvvm.annotations");
+  // The module has no kernels, early exit.
+  if (!AnnotationMD)
+    return;
+
+  for (unsigned I = 0; I < AnnotationMD->getNumOperands(); ++I) {
+    auto Res = getMDNodeAndFuncAtIndex(*AnnotationMD, I);
+    if (!std::get<0>(Res))
+      continue;
+    std::string GlobalName{GlobalNamePrefix};
+    GlobalName.append(std::get<1>(Res)->getName());
+    LLVM_DEBUG(llvm::dbgs() << "Setting up a placeholder for: "
+                            << std::get<1>(Res)->getName() << "\n");
+    // NOTE: The size of the symbol here - 1 - is important (the value is
+    // invalid and would result in a failure from `cuMemcpyHtoD`), instead,
+    // it's used as a flag to communicate to the plugin that even though the
+    // implicit kernel argument is present, it has no uses and hence there is
+    // no need to set up the symbol. See:
+    // `cuda_piextProgramSetSpecializationConstant` for pi handling of it.
+    createGlobal(M, 1, GlobalName, false);
+  }
+}
+
+void CUDASpecConstantToSymbolPass::allocatePerKernelGlobals(NamedMDNode *MD) {
+  Module &M = *MD->getParent();
+  // Calculate the size of global symbol and allocate it for each kernel.
+  for (const auto *Node : MD->operands()) {
+    const StringRef KernelName =
+        cast<MDString>(Node->getOperand(0).get())->getString();
+    LLVM_DEBUG(llvm::dbgs() << "Working on: " << KernelName << "\n");
+    unsigned PerKernelSize = 0;
+    // Loop over all spec constants of a kernel
+    for (unsigned i = 1; i < Node->getNumOperands(); ++i) {
+      MDNode *SC = dyn_cast<MDNode>(Node->getOperand(i));
+      assert(SC && SC->getNumOperands() >= 4 && "Invalid node.");
+      // get the size and offset node to calculate the total size of spec
+      // constants (size of type + offset in the composite - if any);
+      PerKernelSize += uintFromMDNode(SC, SC->getNumOperands() - 1);
+      PerKernelSize += uintFromMDNode(SC, SC->getNumOperands() - 2);
+      LLVM_DEBUG(llvm::dbgs()
+                 << "\tper kernel size calculation: " << PerKernelSize << "\n");
+    }
+    // finally, allocate per-kernel globals.
+    std::string GlobalName{GlobalNamePrefix};
+    GlobalName.append(KernelName.str());
+    SpecConstGlobals[KernelName] =
+        createGlobal(M, PerKernelSize, GlobalName, true);
+  }
+}
+
+void CUDASpecConstantToSymbolPass::fixupSpecConstantUses(
+    Argument *A, const StringRef KernelName) {
+  // Re-write uses of, now defunct, spec constant kernel arg to GEPs from the
+  // global value.
+  IRBuilder B(A->getParent()->getParent()->getContext());
+  SmallVector<Instruction *> ToErase;
+  for (auto *U : A->users()) {
+    auto *I = dyn_cast_or_null<Instruction>(&*U);
```

Comment:
Do we expect `&*U` to result in `nullptr`? Because if not, there is no reason to use `dyn_cast_or_null` and we should just use regular `dyn_cast`.

Moreover, if there is no runtime check for `I != nullptr`, then a regular `cast` should be used

Reply:
The honest answer is that I don't know, it boils down to how the use list is constructed, and it is not clear from the code, it calls `addUse` (https://github.com/jchlanda/llvm/blob/jakub/cuda_native_symbols/llvm/include/llvm/IR/Value.h#L
505) without validating what the `Use` is.

For what it's worth, a quick grep reveals that the code base has all sorts of approaches:
* `dyn_cast_or_null`: https://github.com/jchlanda/llvm/blob/jakub/cuda_native_symbols/llvm/lib/Transforms/IPO/FunctionSpecialization.cpp#L598 (in the call to `getUserBonus`)
* `dyn_cast`: https://github.com/jchlanda/llvm/blob/jakub/cuda_native_symbols/llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp#L2599 (in the call to `classifyArgUse`)

and plenty of straight up accesses to the casted instance.

I guess, we don't expect it to ever by `null`, but have that sanity check `assert(I && "Expected an instruction.");`, will change to `dyn_cast`.